### ISSUE TITLE
Scenario execution priority attributes

### ISSUE THREAD

**Issue by 50U10FCA7** [cucumber-rs/cucumber#187]
Priority attribute tells the executor how early need to add scenario to the execution queue.

In some cases there is no way to tell the executor to run scenarios in specific order. 
For example: Scenarios that will be executing for a long time must be added to execution queue before others.

Design proposal:

```
@priority [integer]
Scenario: my slow scenario
```

Scenarios with lower priority value will be added to execution queue earlier (or vice versa).

**Comment by ilslv** [comment]
@50U10FCA7 @tyranron 

> For example: Scenarios that will be executing for a long time must be added to execution queue before others.

To really solve this issue, in addition to the `@priority=n` tag, we improve default `Runner`. For now, in case of concurrent steps, it collects batch of the `Step`s to execute and waits for until __all__ are complete. We should start executing next `Step` right after any of them is finished. 
Also maybe we should leave an option to wait for all `Step`s in the batch, as it may help with distributing the load on tested application.

**Comment by tyranron** [comment]
@ilslv I'm not sure this is the responsibility of a `Runner`. It seems that a `Parser` may decide which `Features` should be fed first to the `Runner` (however, `Runner` still should be able to distinguish that on `Scenario` level). 

This priority issue doesn't have high priority for us at the moment 😊  
We can give it some time and think about better designs.

> We should start executing next `Step` right after any of them is finished.

That would be worthful improvement anyway.

> Also maybe we should leave an option to wait for all `Step`s in the batch, as it may help with distributing the load on tested application.

How it should help? I don't really grasp your idea.

**Comment by ilslv** [comment]
@tyranron 

> How it should help? I don't really grasp your idea.

If in the batch there is `Scenario` which is very heavy to execute and other concurrent `Scenario`s may throttle it, our batch approach may be helpful, as batch will shrink over time and lower load on the tested target. But maybe it's better to use `@serial` in such cases.

**Comment by tyranron** [comment]
@ilslv yup, or even lower/higher `@priority`.

**Comment by Solonotix** [comment]
Note: I'm an outsider that's new to the Rust community. I just stumbled on this project because I wanted to potentially explore Rust-Cucumber as an option.

Wouldn't priority of scenario be handled by the default parsing behavior of Gherkin files? Maybe this project is different, but the default behavior is to sort steps by the feature file name (maybe file path), then the top-to-bottom order of scenarios. If you want to enforce a priority order, preface the features with a numeral that represents your desired order, and you can even use a dense-ranking system for scenarios that don't need to be run before or after a particular group.

Example:
```
\---features
        01-runs-first.feature
        02-runs-next.feature
        03-indeterminate-test.feature
        03-still-dont-care.feature
        03-technically-fifth.feature
        99-must-run-last.feature
```

**Comment by tyranron** [comment]
@Solonotix this library works the following way:

1. The whole execution pipeline consists of [`Parser` -> `Runner` -> `Writer`s](https://cucumber-rs.github.io/cucumber/current/architecture/index.html).
2. The parser of Gherkin files is abstracted in the way to be just a `Stream` of parsed `Feature`s. It could be anything emitting `Feature`s, not just glob-walker over fs `.feature` files.
3. `Runner` polls the `Parser`'s `Stream` and executes `Scenario`s concurrently (the concurrency level [can be configured](https://cucumber-rs.github.io/cucumber/current/writing/tags.html#isolated-execution)).
4. `Writer` receives events about execution status from `Runner` in out-of-order manner, but the `writer::Normalized` normalizes the output in the way that once some `Feature` is finished, it's fully printed right away.

So, at the moment, we're not preserving any order when executing and printing `Scenario`s, only steps inside a `Scenario` are executed sequentially.

**Comment by ilslv** [comment]
@Solonotix I also would argue that `Scenario`s should be distributed across `.feature`-files based on some [domain-specific needs][1], not execution priority. 




[1]: https://cucumber-rs.github.io/cucumber/current/writing/modules.html#grouping
